[ES|QL] Add support for STATS BY nested function

dej611 opened the pull request:
## Summary

Add support for the validation and autocomplete engine for the feature introduced in https://github.com/elastic/elasticsearch/pull/104387

The validation engine should not mark the syntax as invalid now.
The autocomplete changes are a little bit more subtle:
* after the `by` option command a `varX` and functions are now suggested in additions to `[columns]`
* when typing an expression within the `by` scope the autocomplete should now understand and help with that, without promoting it when not needed.

![by_suggestions](https://github.com/elastic/kibana/assets/924948/266c403b-34a6-436f-a5aa-43353b77269e)


### Checklist

- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios

dej611 commented:
/ci

dej611 commented:
/ci

dej611 commented:
/ci

elasticmachine commented:
Pinging @elastic/kibana-visualizations (Team:Visualizations)

stratoula commented:
/ci

stratoula commented:
@dej611 as this is a very cool feature, can we suggest also all the eval functions (as they are supported). 

<img width="1578" alt="image" src="https://github.com/elastic/kibana/assets/17003240/4cc7349a-74c1-4d75-9482-ffe534d1c5d4">

dej611 commented:
I wasn't sure whether to suggest them or not, but I can enable that.
I was just concerned by the suggestion long list in this context.
To be sure, this is what I can do:
* suggest a function after the `by`
* suggest a function after the `=` (assign) operator within the `by` scope

Does it make sense to you @stratoula ?

stratoula commented:
Yes!!! I think is cool to suggest them even if the list goes long!

dej611 commented:
/ci

dej611 commented:
Updated logic to propose also functions.
Added also additional tests for more scenarios.

![by_suggestions](https://github.com/elastic/kibana/assets/924948/4817e89b-d2c0-4812-8a3b-157475d44c56)

stratoula reviewed (approved):
This looks great Marco! LGTM, thanx a ton

dej611 commented:
I've completely missed the support for nested functions in the `STATS` scope in this PR.
I'll make a follow up PR with the support for that.
